### ISSUE TITLE
Documentation to use torchgeo to create tiles from large rasters

### ISSUE THREAD

**Issue by RitwikGupta** [microsoft/torchgeo#280]
To the best of my knowledge, I am unable to find documentation in torchgeo that describes how I can sample `N x N` tiles from a custom Dataset which loads full rasters from the image.

This sampling workflow would ideally pad my returned raster such that it results in evenly sized tiles.

**Comment by adamjstewart** [comment]
Can you clarify what you're trying to do? When you say "custom Dataset", are you talking about a non-torchgeo dataset? Is it possible to implement that as a torchgeo GeoDataset? If it can be implemented as a GeoDataset, it should be easy to use any of our GeoSamplers to sample tiles of fixed size.

**Comment by RitwikGupta** [comment]
Sorry, I should say custom `GeoDataset` or `RasterDataset`. I believe the issue I was having with these Samplers is that they all require some `dataset.index` to be created, which from other models seem to involve bounding boxes. I have no labels, I want to just sample tiles. I also have no idea how to create this `dataset.index`.

**Comment by adamjstewart** [comment]
Ah, I see. The feature you're interested in (sampling tiles from a GeoDataset) is already present. What you really want is documentation on how to write a custom GeoDataset. This is definitely something I want to add better documentation for. 

Basically, every GeoDataset has a `dataset.index` object. This object is an R-tree containing the bounding boxes and filenames of all raster or vector files in the dataset. These files can either be input images or target labels. Your best bet is to look at the implementations of other GeoDataset subclasses (anything in [this list](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#geospatial-datasets)). If you subclass RasterDataset, all you need to do is provide any globs/regexes that explain how to parse the filenames for information like the timestamp of the image. If you don't care about the timestamp (all images come from the same time or time doesn't matter), you don't need to override anything (CDL is a good example of this).

Which dataset are you trying to add? If you give me a download link or some example filenames I can help write the globs/regexes you'll need. If it's a public dataset, we would love to add it to TorchGeo!

**Comment by RitwikGupta** [comment]
The dataset I am working with is currently not public. This gist, however, demonstrates that there is some error even when following something similar to CDL: https://gist.github.com/RitwikGupta/334807f1f081ed46686613016c265205

**Comment by isaaccorley** [comment]
@RitwikGupta Regarding your gist example, can you try calling super() at the top of your constructor and not set any of the attributes. I think your issue is coming from you setting self.crs before superclassing the parent classes. In other words, we have a setter method for crs which expects a _crs attribute to exist but this is not created until the RasterDataset constructor is called via super().

**Comment by RitwikGupta** [comment]
That still results in an error related to `dataset.index`: https://gist.github.com/RitwikGupta/f4cfe9370b444b435181e4c15776b68a

**Comment by adamjstewart** [comment]
The default sampler for a PyTorch `DataLoader` uses integers to query the dataset. However, integer indexing for `GeoDatasets` doesn't make sense. You'll need to use one of our `GeoSamplers` instead. See the README or tutorial for an example of this.

**Comment by RitwikGupta** [comment]
Cool. I have a working demo of a custom `RasterDataset`. Where should I push a working tutorial?

**Comment by ashnair1** [comment]
Tutorials are stored under docs [here](https://github.com/microsoft/torchgeo/tree/main/docs/tutorials)
